### ISSUE TITLE
Transformed RVs passed into other RVs do not work properly

### ISSUE THREAD

**Issue by twiecki** [pymc-devs/pymc4#89]
```python
@pm.model
def t_test(sd_prior='half_normal'):
    mu = pm.Normal('mu', 0, 1)
    sd = pm.HalfNormal('sd', 1)
    # Problem here! sd still can't take negative values when passed into other vars
    pm.Normal('y_0', 0, 2 * sd)
    pm.Normal('y_1', mu, 2 * sd)

model = t_test.configure()

model._forward_context.vars
```

```python
func = model.make_log_prob_function()

mu = -tf.ones((10,))
sd = -tf.ones((10,))
y_0 = tf.ones((10,))
y_1 = tf.ones((10,))

func(mu, sd, y_0, y_1)
```

Returns nan. The problem is actually not in line 2 of the model, but when it is passed on in line 3. So what we need to do is back-transform the value for when it's seen by other RVs. Not quite sure where we need to do that but maybe we can look at how PyMC3 does it.

CC @canyon289 @ericmjl

**Comment by junpenglao** [comment]
Hmm, I think we should think more carefully here - I would expect that if you are passing a value to `sd` it should be positive only. In PyMC3 we make the distinction in 
1. the `point` that we pass to `logp`
2. the transformed RV being something like `sd_log__` that is actually in the graph and the original `sd` being a deterministic.

Which means, when user pass value for logp evaluation, there should be a way to distinguish whether the intention is the transformed or non-transformed RV.

**Comment by ericmjl** [comment]
@twiecki, are you sure your example is correct? In `sd`, you passed in negative signed ones.

**Comment by ericmjl** [comment]
I did a chat with @canyon289 this evening to talk through the problem. I created a smaller, more minimal version of the issue.

```python
import tensorflow as tf
import pymc4 as pm

session = tf.Session()

@pm.model
def model_func():
    sigma = pm.HalfNormal('sigma', 1)
    data = pm.Normal('data', mu=0, sigma=sigma)

model = model_func.configure()

func = model.make_log_prob_function()

sigma = -tf.ones(1)
data = tf.ones(1)

sess = tf.Session()
with sess:
    print(func(sigma, data).eval())  # nan
```

First question. In `model_func()`, when `sigma` is passed into `pm.Normal`, is a number being passed to it, or is it something else?

Second question. What is the data path for the `tf.ones` when they are passed into the log probability function? 

I suspect we need knowledge of the context (forward, free forward, and inference) to know what's going on?

**Comment by twiecki** [comment]
I think I figured it out. Transformations can easily be done manually as part of the model definition. E.g.
```
log(tau) ~ Normal(0, 1) # we do this correctly
x ~ Normal(0, exp(tau)) # we don't do this
```
So what we need to do is transform the RV, just like we would do manually. There is some oddities in how the stuff works with all the contexts but briefly, RVs register themselves in the context (forward-context at first). Then when actually building the graph we retrieve the RV from the context and plug everything together.

I gave it a try here: https://github.com/pymc-devs/pymc4/pull/90 which seems to work.
